internal/cli/clients/push.go

```diff
@@ -131,53 +125,25 @@ func (c *EpinioClient) deployCode(req models.DeployRequest) ([]byte, error) {
 		return nil, errors.Wrap(err, "can't deploy app")
 	}
 
-	return b, nil
+	// returns app default route
+	deploy := &models.DeployResponse{}
+	if err := json.Unmarshal(b, deploy); err != nil {
+		return nil, err
+	}
+
+	return deploy, nil
 }
 
 func (c *EpinioClient) waitForPipelineRun(ctx context.Context, app models.AppRef, id string) error {
 	c.ui.ProgressNote().KeeplineUnder(1).Msg("Running staging")
 
-	cs, err := tekton.NewForConfig(c.Cluster.RestConfig)
-	if err != nil {
-		return err
-	}
-	client := cs.TektonV1beta1().PipelineRuns(deployments.TektonStagingNamespace)
-
-	return wait.PollImmediate(time.Second, duration.ToAppBuilt(),
-		func() (bool, error) {
-			l, err := client.List(ctx, metav1.ListOptions{LabelSelector: models.EpinioStageIDLabel + "=" + id})
-			if err != nil {
-				return false, err
-			}
-			if len(l.Items) == 0 {
-				return false, nil
-			}
-			for _, pr := range l.Items {
-				// any failed conditions, throw an error so we can exit early
-				for _, c := range pr.Status.Conditions {
-					if c.IsFalse() {
-						return false, errors.New(c.Message)
-					}
-				}
-				// it worked
-				if pr.Status.CompletionTime != nil {
-					return true, nil
-				}
-			}
-			// pr exists, but still running
-			return false, nil
-		})
+	_, err := c.get(api.Routes.Path("StagingComplete", app.Org, id))
```

Comment:
I'm wondering if the request timeout is long enough for the server side timeout (ToAppBuilt) to be effective. I mean, we don't set any timeout here but that doesn't mean the cli won't timeout before the server does. If that is true, we no longer control how long the cli waits for staging to complete. Maybe we need to pass something to the http client to achieve that.

Reply:
Hm. Will check. If true will put in some more thought.

Reply:
The server side timeout are handled at

- https://github.com/epinio/epinio/blob/fc07dc09a1603908ac1351925e54a83bc62145b6/internal/api/v1/stage.go#L260
- https://github.com/epinio/epinio/blob/fc07dc09a1603908ac1351925e54a83bc62145b6/internal/api/v1/applications.go#L333

It is (default) 10 minutes.

Where is the client timeout handled ... Oh, wait, you are saying above that the client does not have a timeout here ...
While I could see a browser timing out, why do you believe the client would despite not having set one ?
AFAICT from the net/http docs the default timeout is `no timeout`.

Reply:
ok then